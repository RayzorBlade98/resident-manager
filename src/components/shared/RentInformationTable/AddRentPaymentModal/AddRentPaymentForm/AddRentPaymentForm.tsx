import { Grid } from '@mui/material';
import React from 'react';
import { useRecoilState } from 'recoil';
import CurrencyInputField from '_/components/form/CurrencyInputField/CurrencyInputField';
import StandardDateField from '_/components/form/StandardDateField/StandardDateField';
import {
  addRentPaymentFormValidationSelector,
  RentPaymentInput,
} from '_/components/shared/RentInformationTable/states/add_rent_payment_state';

/**
 * Form to submit payment information
 */
function AddRentPaymentForm(): JSX.Element {
  const [formValidationState, setFormValidationState] = useRecoilState(
    addRentPaymentFormValidationSelector,
  );
  const formInput = formValidationState.formInput;
  const errors = formValidationState.formErrors;

  function onChange<T>(field: keyof RentPaymentInput, value: T): void {
    setFormValidationState((state) => ({
      ...state,
      formInput: {
        ...state.formInput,
        [field]: value,
      },
    }));
  }

  return (
    <form>
      <Grid container columnSpacing={2} rowSpacing={2}>
        <Grid item xs={12}>
          <CurrencyInputField
            required
            id="paymentAmount"
            label="Zahlungssumme"
            value={formInput.paymentAmount}
            onChange={(value) => {
              onChange<number | undefined>('paymentAmount', value);
            }}
            errorMessage={errors.paymentAmount}
          />
        </Grid>
        <Grid item xs={12}>
          <StandardDateField
            required
            id="paymentDate"
            label="Zahlungsempfang"
            value={formInput.paymentDate}
            onChange={(value) => {
              onChange<Date | undefined>('paymentDate', value);
            }}
            errorMessage={errors.paymentDate}
          />
        </Grid>
      </Grid>
    </form>
  );
}

export default AddRentPaymentForm;
