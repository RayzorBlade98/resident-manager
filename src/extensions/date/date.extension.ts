declare interface Date {
  /**
   * Converts the date to a new date in UTC timezone
   */
  toUTC(): Date;

  /**
   * Converts the date to 'DD.MM.YYYY' format string
   */
  toPreferredString(): string;
}

function toUTC(this: Date): Date {
  return new Date(
    Date.UTC(this.getFullYear(), this.getMonth(), this.getDate(), 0, 0, 0, 0),
  );
}
Date.prototype.toUTC = toUTC;

function toPreferredString(this: Date): string {
  const day: number = this.getDate();
  const month: number = this.getMonth() + 1;
  const year = String(this.getFullYear());

  const dayPrefix: string = day < 10 ? '0' : '';
  const monthPrefix: string = month < 10 ? '0' : '';

  return `${dayPrefix}${String(day)}.${monthPrefix}${String(month)}.${year}`;
}
Date.prototype.toPreferredString = toPreferredString;
