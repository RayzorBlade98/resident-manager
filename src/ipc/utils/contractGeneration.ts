import {
  convertAddressToCityString,
  convertAddressToStreetString,
} from '../../utils/address/address.utils';
import { convertNameToString } from '../../utils/name/name.utils';
import contractTemplate from '_/assets/contract/contractTemplate.md';
import landlordCompanyTemplate from '_/assets/contract/landlordCompanyTemplate.md';
import Landlord from '_/models/landlord/landlord';
import Property from '_/models/property/property';
import { Resident } from '_/models/resident/resident';

export type ContractGenerationArgs = {
  landlord: Landlord;
  resident: Resident;
  property: Property;
};

const placeholderLabels = {
  landlordName: 'LANDLORD_NAME',
  landlordStreet: 'LANDLORD_STREET',
  landlordCity: 'LANDLORD_CITY',
  landlordPhone: 'LANDLORD_PHONE', // Todo
  landlordEmail: 'LANDLORD_EMAIL', // Todo
  landlordCompany: 'LANDLORD_COMPANY',
  residentName: 'RESIDENT_NAME',
  residentStreet: 'RESIDENT_STREET', // Todo
  residentCity: 'RESIDENT_CITY', // Todo
  residentPhone: 'RESIDENT_PHONE', // Todo
  propertyStreet: 'PROPERTY_STREET',
  propertyCity: 'PROPERTY_CITY',
  apartmentFloor: 'APARTMENT_FLOOR',
  apartmentLocation: 'APARTMENT_LOCATION',
  apartmentRoomsGeneric: 'APARTMENT_ROOMS_GENERIC',
  apartmentRoomsKitchen: 'APARTMENT_ROOMS_KITCHEN',
  apartmentRoomsBath: 'APARTMENT_ROOMS_BATH',
  apartmentRoomsHallway: 'APARTMENT_ROOMS_HALLWAY',
  apartmentRoomsBasement: 'APARTMENT_ROOMS_BASEMENT',
  apartmentRoomsGarden: 'APARTMENT_ROOMS_GARDEN', // Todo
  parkingSpaceCount: 'PARKING_SPACE_COUNT',
  keysApartment: 'KEYS_APARTMENT',
  keysBasement: 'KEYS_BASEMENT',
  keysAttic: 'KEYS_ATTIC',
  keysFrontdoor: 'KEYS_FRONTDOOR',
  keysMailbox: 'KEYS_MAILBOX',
} satisfies Record<string, string>;

const blockPlaceholderLabels = {
  landlordCompany: 'LANDLORD_COMPANY_BLOCK',
} satisfies Record<string, string>;

/**
 * Generates the contract as markdown string
 * @returns markdown string of the generated contract
 */
export function generateContractMarkdown(args: ContractGenerationArgs): string {
  const apartment = args.property.apartments.find(
    (a) => a.id === args.resident.apartmentId,
  )!;

  const placeholders = {
    [placeholderLabels.landlordName]: convertNameToString(
      args.landlord.representative,
    ),
    [placeholderLabels.landlordStreet]: convertAddressToStreetString(
      args.landlord.address,
    ),
    [placeholderLabels.landlordCity]: convertAddressToCityString(
      args.landlord.address,
    ),
    [placeholderLabels.landlordCompany]: args.landlord.company ?? '',
    [placeholderLabels.residentName]: convertNameToString(args.resident.name),
    [placeholderLabels.propertyStreet]: convertAddressToStreetString(
      args.property.address,
    ),
    [placeholderLabels.propertyCity]: convertAddressToCityString(
      args.property.address,
    ),
    [placeholderLabels.apartmentFloor]: apartment.floor,
    [placeholderLabels.apartmentLocation]: apartment.location,
    [placeholderLabels.apartmentRoomsGeneric]:
      apartment.rooms.generic.toString(),
    [placeholderLabels.apartmentRoomsKitchen]:
      apartment.rooms.kitchen.toString(),
    [placeholderLabels.apartmentRoomsBath]: apartment.rooms.bath.toString(),
    [placeholderLabels.apartmentRoomsBasement]:
      apartment.rooms.basement.toString(),
    [placeholderLabels.apartmentRoomsHallway]:
      apartment.rooms.hallway.toString(),
    [placeholderLabels.parkingSpaceCount]: args.resident.parkingSpaceId ? '1' : '0',
    [placeholderLabels.keysApartment]: args.resident.keys.apartment.toString(),
    [placeholderLabels.keysBasement]: args.resident.keys.basement.toString(),
    [placeholderLabels.keysAttic]: args.resident.keys.attic.toString(),
    [placeholderLabels.keysFrontdoor]: args.resident.keys.frontDoor.toString(),
    [placeholderLabels.keysMailbox]: args.resident.keys.mailbox.toString(),
  };

  let contract = contractTemplate;

  // Landlord company
  contract = contract.replace(
    getPlaceholder(blockPlaceholderLabels.landlordCompany),
    args.landlord.company ? landlordCompanyTemplate : '',
  );

  return replaceAllPlaceholders(contract, placeholders);
}

function replaceAllPlaceholders(
  contract: string,
  replacements: Record<string, string>,
): string {
  Object.entries(replacements).forEach(([label, replacement]) => {
    contract = contract.replace(getPlaceholder(label), replacement);
  });
  return contract;
}

function getPlaceholder(label: string): string {
  return `{{${label}}}`;
}
