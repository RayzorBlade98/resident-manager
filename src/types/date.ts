/**
 * Enum containing all months as `string`
 */
export enum Month {
  January = 'Januar',
  Febuary = 'Februar',
  March = 'MÃ¤rz',
  April = 'April',
  May = 'Mai',
  June = 'Juni',
  July = 'Juli',
  August = 'August',
  September = 'September',
  October = 'Oktober',
  November = 'November',
  December = 'Dezember',
}

/**
 * Type representing a year
 */
export type Year = number;

/**
 * Information about a specific month and year
 */
export interface MonthYear {
  month: Month;
  year: Year;
}

/**
 * Type of the string representation of a `MonthYear`
 *
 * Format: "[month] [year]"
 */
export type MonthYearString = `${Month} ${Year}`;

/**
 * Type of the string representation of a date including day, month and year
 *
 * Format: `dd.mm.yyyy`
 */
export type DateString = `${string}.${string}.${string}`;

/**
 * Type of the date string representation returned by the date input field
 *
 * Format: `yyyy-mm-dd`
 */
export type DateInputString = `${string}-${string}-${string}`;

/**
 * Class that provides utility functions for `MonthYear` objects
 */
export abstract class MonthYearUtils {
  /**
   * Adds the specified number of months to the given `MonthYear` object
   * @param monthYear `MonthYear` object to which the months should be added
   * @param addedMonths Number of months that should be added (can be negative)
   * @returns new `MonthYear` object with the adjusted month (and year)
   */
  public static addMonths(monthYear: MonthYear, addedMonths = 1): MonthYear {
    if (addedMonths < 0) {
      // Add negative amount of months == subtract months
      return MonthYearUtils.subtractMonths(monthYear, addedMonths * -1);
    }

    const newMonthYear = { ...monthYear };
    if (newMonthYear.month === Month.December) {
      // Year change
      newMonthYear.month = Month.January;
      newMonthYear.year += 1;
    } else {
      // Add one month to the object
      const months = Object.values(Month);
      const oldMonthIndex = months.findIndex(
        (m: Month) => m === newMonthYear.month,
      );
      newMonthYear.month = months[oldMonthIndex + 1];
    }

    // Call `addMonths` recursive until all months got addded
    if (addedMonths !== 1) {
      return MonthYearUtils.addMonths(newMonthYear, addedMonths - 1);
    }
    return newMonthYear;
  }

  /**
   * Checks if two `MonthYear` objects are equal
   * @param a first `MonthYear` object to compare
   * @param b second `MonthYear` object to compare
   * @returns `true` if both the month and year are the same, else `false`
   */
  public static areEqual(a: MonthYear, b: MonthYear): boolean {
    return a.month === b.month && a.year === b.year;
  }

  /**
   * Compares two `MonthYear` objects
   * @param a First `MonthYear` object to compare
   * @param b Second `MonthYear` object to compare
   * @returns
   *
   * - 0 if both object have the same month and year
   * - <0 if the first object contains a date that's earlier than the second object
   * - \>0 if the first object contains a date that's later than the second object
   */
  public static compare(a: MonthYear, b: MonthYear): number {
    if (a.year !== b.year) return a.year - b.year;

    const months = Object.values(Month);
    const aMonthIndex = months.findIndex((m: Month) => m === a.month);
    const bMonthIndex = months.findIndex((m: Month) => m === b.month);
    return aMonthIndex - bMonthIndex;
  }

  /**
   * Returns a `MonthYear` object that contains information of the current month and year
   * @returns `MonthYear` object for the current date
   */
  public static getCurrentMonthYear(): MonthYear {
    const date: Date = new Date();
    const monthIndex: number = date.getUTCMonth();
    const month: Month = Object.values(Month)[monthIndex] as Month;
    const year: Year = date.getUTCFullYear();

    return {
      month,
      year,
    };
  }

  /**
   * Parses a string representation of a `MonthYear` object to its respective object
   * @param monthYear string representation of a `MonthYear` object that should be parsed
   * @returns `MonthYear` object that contains the information of the specified `string`
   */
  public static parseString(monthYear: MonthYearString): MonthYear {
    const splits = monthYear
      .split(/(\s+)/)
      .filter((s: string) => s.trim().length > 0);
    return {
      month: splits[0] as Month,
      year: Number(splits[1]),
    };
  }

  /**
   * Subtracts the specified number of months from the given `MonthYear` object
   * @param monthYear `MonthYear` object from which the months should be subtraced
   * @param subtractedMonths Number of months that should be subtracted (can be negative)
   * @returns new `MonthYear` object with the adjusted month (and year)
   */
  public static subtractMonths(
    monthYear: MonthYear,
    subtractedMonths = 1,
  ): MonthYear {
    if (subtractedMonths < 0) {
      // Subtract negative amount of months == add months
      return MonthYearUtils.addMonths(monthYear, subtractedMonths * -1);
    }

    const newMonthYear = { ...monthYear };
    if (newMonthYear.month === Month.January) {
      // Year change
      newMonthYear.month = Month.December;
      newMonthYear.year -= 1;
    } else {
      // Subtract one month from the object
      const months = Object.values(Month);
      const oldMonthIndex = months.findIndex(
        (m: Month) => m === newMonthYear.month,
      );
      newMonthYear.month = months[oldMonthIndex - 1];
    }

    // Call `subtractMonths` recursive until all months got subtracted
    if (subtractedMonths !== 1) {
      return MonthYearUtils.subtractMonths(newMonthYear, subtractedMonths - 1);
    }
    return newMonthYear;
  }

  /**
   * Converts a `MonthYear` to its string representation
   * @param monthYear `MonthYear` that should be converted
   * @returns string representation of the specified `MonthYear` object with the format "[month] [year]"
   */
  public static toString(monthYear: MonthYear): MonthYearString {
    return `${monthYear.month} ${monthYear.year}`;
  }

  /**
   * Creates a list of `MonthYear` objects that contains each month
   * between the two specified `MonthYear` objects (inclusive these months)
   * @param start Start of the timespan
   * @param end End of the timespan
   * @returns List of `MonthYear` objects with all months between `start` and `end`
   */
  public static timespan(start: MonthYear, end: MonthYear): MonthYear[] {
    const timespan = [{ ...start }];

    // Whether to add a month each loop or subtract one
    const direction = Math.sign(MonthYearUtils.compare(end, start));

    if (MonthYearUtils.areEqual(start, end)) return timespan;

    for (
      let next = MonthYearUtils.addMonths(start, direction);
      !MonthYearUtils.areEqual(next, end);
      next = MonthYearUtils.addMonths(next, direction)
    ) {
      timespan.push(next);
    }
    timespan.push({ ...end });
    return timespan;
  }
}

/**
 * Class that provides utility functions for dates containing day, month and year.
 */
export abstract class DateUtils {
  /**
   * Converts a date string from `yyyy-mm-dd` to `dd.mm.yyyy` format
   * @param date Date in `yyyy-mm-dd` format
   * @returns Converted date in `dd.mm.yyyy` format
   */
  public static convertDateInputString(date: DateInputString): DateString {
    const splits = date.split('-');
    return `${splits[2]}.${splits[1]}.${splits[0]}`;
  }

  /**
   * Converts a date string from `dd.mm.yyyy` to `yyyy-mm-dd` format
   * @param date Date in `dd.mm.yyyy` format
   * @returns Converted date in `yyyy-mm-dd` format
   */
  public static convertToDateInputString(date: DateString): DateInputString {
    const splits = date.split('.');
    return `${splits[2]}-${splits[1]}-${splits[0]}`;
  }

  /**
   * Returns a date string containing the current date
   * @returns Date string containing the current date in `dd.mm.yyyy` format
   */
  public static getCurrentDate(): DateString {
    const date: Date = new Date();
    const day: number = date.getUTCDate();
    const month: number = date.getUTCMonth() + 1;
    const year = String(date.getUTCFullYear());

    const dayPrefix: string = day < 10 ? '0' : '';
    const monthPrefix: string = month < 10 ? '0' : '';

    return `${dayPrefix}${String(day)}.${monthPrefix}${String(month)}.${year}`;
  }
}
