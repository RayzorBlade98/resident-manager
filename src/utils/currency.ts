export function convertIntToCurrency(int: number): string {
  const float = (int / 100.0).toFixed(2);
  return `${float}â‚¬`;
}

export function convertCurrencyFloatToInt(float: number): number {
  return float * 100;
}
