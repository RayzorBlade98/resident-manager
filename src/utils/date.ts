// Remove after adding another export:
/* eslint-disable import/prefer-default-export */

/**
 * Converts the given date to its string representaion (format: DD.MM.YYYY)
 * @param date date that should be converted
 * @returns string representation of the date
 */
export function dateToString(date: Date): string {
  const day: number = date.getDate();
  const month: number = date.getMonth() + 1;
  const year = String(date.getFullYear());

  const dayPrefix: string = day < 10 ? '0' : '';
  const monthPrefix: string = month < 10 ? '0' : '';

  return `${dayPrefix}${String(day)}.${monthPrefix}${String(month)}.${year}`;
}
