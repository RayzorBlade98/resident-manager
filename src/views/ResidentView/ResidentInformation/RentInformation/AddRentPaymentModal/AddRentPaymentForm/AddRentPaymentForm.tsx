import { Grid } from '@mui/material';
import React from 'react';
import { useRecoilState, useRecoilValue } from 'recoil';
import {
  addRentPaymentFormErrorSelector,
  addRentPaymentFormInputSelector,
} from '../../states/add_rent_payment_state';
import CurrencyInputField from '_/components/CurrencyInputField/CurrencyInputField';
import StandardDateField from '_/components/StandardDateField/StandardDateField';

/**
 * Form to submit payment information
 */
function AddRentPaymentForm(): JSX.Element {
  const [formInput, setFormInput] = useRecoilState(
    addRentPaymentFormInputSelector,
  );
  const errors = useRecoilValue(addRentPaymentFormErrorSelector);

  const onChangePaymentAmount = (value: number | undefined) => {
    setFormInput((input) => ({ ...input, paymentAmount: value }));
  };

  const onChangePaymentDate = (date: Date | undefined) => {
    setFormInput((input) => ({ ...input, paymentDate: date }));
  };

  return (
    <form>
      <Grid container columnSpacing={2} rowSpacing={2}>
        <Grid item xs={12}>
          <CurrencyInputField
            required
            id="paymentAmount"
            label="Zahlungssumme"
            value={formInput.paymentAmount}
            onChange={onChangePaymentAmount}
            errorMessage={errors.paymentAmount}
          />
        </Grid>
        <Grid item xs={12}>
          <StandardDateField
            required
            id="paymentDate"
            label="Zahlungsempfang"
            value={formInput.paymentDate}
            onChange={onChangePaymentDate}
            errorMessage={errors.paymentDate}
          />
        </Grid>
      </Grid>
    </form>
  );
}

export default AddRentPaymentForm;
