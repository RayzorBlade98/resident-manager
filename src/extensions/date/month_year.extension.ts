import './date.extension';

const MONTH_NAMES = [
  'Januar',
  'Februar',
  'MÃ¤rz',
  'April',
  'Mai',
  'Juni',
  'Juli',
  'August',
  'September',
  'Oktober',
  'November',
  'Dezember',
];

/**
 * Date object that represents a month and a year
 */
class MonthYear extends Date {
  /**
   * Creates a new `MonthYear` object
   * @param month month of the new object. If undefined the current month will be used.
   * @param year year of the new object. If undefined the current year will be used.
   */
  constructor(month?: number, year?: number) {
    const now = new Date();
    const utc = new Date(
      year !== undefined ? year : now.getFullYear(),
      month !== undefined ? month : now.getMonth(),
      1,
    ).toUTC();
    super(utc.getTime());
  }

  /**
   * Converts a `Date` object to a `MonthYear`
   * @param date Date that should be converted
   * @returns `MonthYear` with the same month and year as `date`
   */
  public static fromDate(date: Date): MonthYear {
    return new MonthYear(date.getMonth(), date.getFullYear());
  }

  /**
   * Converts a date string to a `MonthYear`
   * @param date Date string that should be converted
   * @returns `MonthYear` with the same month and year as `date`
   */
  public static fromString(date: string): MonthYear {
    return MonthYear.fromDate(new Date(date));
  }

  /**
   * Adds the specified number of months to the given `MonthYear` object
   * @param toAdd Number of months that should be added (can be negative)
   * @returns this month year
   */
  public addMonths(toAdd = 1): MonthYear {
    this.setMonth(this.getMonth() + toAdd);
    return this;
  }

  /**
   * Creates a copy this object.
   * @returns the copy with the same month and year
   */
  public clone(): MonthYear {
    return new MonthYear(this.getMonth(), this.getFullYear());
  }

  /**
   * Checks whether this object has the same month and year as anoter `MonthYear`
   * @param other true if both the month and year are the same, else false
   */
  public equals(other: MonthYear): boolean {
    return this >= other && this <= other;
  }

  /**
   * Creates a list of `MonthYear` objects that contains each month
   * between the two specified `MonthYear` objects (inclusive these months)
   * @param start Start of the timespan
   * @param end End of the timespan
   * @returns List of `MonthYear` objects with all months between `start` and `end`
   */
  public static timespan(start: MonthYear, end: MonthYear): MonthYear[] {
    // Whether to add a month each loop or subtract one
    const direction = end > start ? 1 : -1;

    const timespan: MonthYear[] = [start.clone()];
    let next = start;
    while (!end.equals(next)) {
      next = next.clone();
      next.addMonths(direction);
      timespan.push(next);
    }

    return timespan;
  }

  /**
   * Generates a string representation in the format `<month name> <year>`
   * @returns string representation of this object
   */
  public toString(): string {
    return `${MONTH_NAMES[this.getMonth()]} ${this.getFullYear()}`;
  }
}

export default MonthYear;
