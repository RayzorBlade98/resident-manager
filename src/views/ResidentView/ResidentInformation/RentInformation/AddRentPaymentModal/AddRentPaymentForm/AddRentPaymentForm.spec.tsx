import {
  RenderResult, act, fireEvent, render,
} from '@testing-library/react';
import React from 'react';
import { getRecoil, resetRecoil } from 'recoil-nexus';
import addRentPaymentState, {
  addRentPaymentFormValidationSelector,
} from '../../states/add_rent_payment_state';
import AddRentPaymentForm from './AddRentPaymentForm';
import { convertCurrencyEurosToCents } from '_/utils/currency/currency';
import { dateToUTC } from '_/utils/date';
import RecoilTestWrapper from '_tests/__test_utils__/RecoillTestWrapper';

describe('AddRentPaymentForm', () => {
  let renderResult: RenderResult;

  function inputToForm(paymentAmount: number, paymentDate: string) {
    const inputFields = renderResult.container.querySelectorAll('input');
    fireEvent.change(inputFields.item(0), {
      target: { value: `${paymentAmount}` },
    });

    fireEvent.change(inputFields.item(1), {
      target: { value: paymentDate },
    });
  }

  beforeEach(() => {
    renderResult = render(
      <RecoilTestWrapper>
        <AddRentPaymentForm />
      </RecoilTestWrapper>,
    );

    act(() => {
      resetRecoil(addRentPaymentState);
    });
  });

  test('should add input to state', () => {
    // Arrange
    const paymentAmount = 100;

    // Act
    inputToForm(paymentAmount, '25.12.1998');

    // Assert
    const formInput = getRecoil(addRentPaymentFormValidationSelector).formInput;
    expect(formInput).toEqual({
      paymentAmount: convertCurrencyEurosToCents(paymentAmount),
      paymentDate: dateToUTC(new Date(1998, 11, 25)),
    });
  });

  test('should match snapshot (valid input)', () => {
    // Act
    inputToForm(123, '28.05.2023');

    // Assert
    expect(renderResult.container).toMatchSnapshot();
  });

  test('should match snapshot (invalid input)', () => {
    // Act
    inputToForm(-1, '');

    // Assert
    expect(renderResult.container).toMatchSnapshot();
  });
});
