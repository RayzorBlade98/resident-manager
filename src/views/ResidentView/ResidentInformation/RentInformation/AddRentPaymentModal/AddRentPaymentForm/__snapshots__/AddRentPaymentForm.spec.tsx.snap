// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddRentPaymentForm should match snapshot (invalid input) 1`] = `
<div>
  <form>
    <div
      class="MuiGrid-root MuiGrid-container css-mhc70k-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
      >
        <div
          class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-error MuiFormLabel-filled Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="true"
            for="paymentAmount"
            id="paymentAmount-label"
          >
            Zahlungssumme
            <span
              aria-hidden="true"
              class="MuiFormLabel-asterisk MuiInputLabel-asterisk Mui-error css-wgai2y-MuiFormLabel-asterisk"
            >
               
              *
            </span>
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary Mui-error MuiInputBase-formControl MuiInputBase-adornedEnd css-o9k5xi-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-describedby="paymentAmount-helper-text"
              aria-invalid="true"
              class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
              id="paymentAmount"
              required=""
              type="number"
              value="-1"
            />
            <div
              class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 css-1pnmrwp-MuiTypography-root"
              >
                €
              </p>
            </div>
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-14lo706"
              >
                <span>
                  Zahlungssumme
                   
                  *
                </span>
              </legend>
            </fieldset>
          </div>
          <p
            class="MuiFormHelperText-root Mui-error MuiFormHelperText-sizeMedium MuiFormHelperText-contained MuiFormHelperText-filled Mui-required css-1wc848c-MuiFormHelperText-root"
            id="paymentAmount-helper-text"
          >
            Muss größer als 0 sein!
          </p>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
      >
        <div
          class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-error Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for="paymentDate"
            id="paymentDate-label"
          >
            Zahlungsempfang
            <span
              aria-hidden="true"
              class="MuiFormLabel-asterisk MuiInputLabel-asterisk Mui-error css-wgai2y-MuiFormLabel-asterisk"
            >
               
              *
            </span>
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary Mui-error MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-describedby="paymentDate-helper-text"
              aria-invalid="true"
              autocomplete="off"
              class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id="paymentDate"
              inputmode="text"
              placeholder="TT.MM.JJJJ"
              required=""
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-yjsfm1"
              >
                <span>
                  Zahlungsempfang
                   
                  *
                </span>
              </legend>
            </fieldset>
          </div>
          <p
            class="MuiFormHelperText-root Mui-error MuiFormHelperText-sizeMedium MuiFormHelperText-contained Mui-required css-1wc848c-MuiFormHelperText-root"
            id="paymentDate-helper-text"
          >
            Darf nicht leer sein!
          </p>
        </div>
      </div>
    </div>
  </form>
</div>
`;

exports[`AddRentPaymentForm should match snapshot (valid input) 1`] = `
<div>
  <form>
    <div
      class="MuiGrid-root MuiGrid-container css-mhc70k-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
      >
        <div
          class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="true"
            for="paymentAmount"
            id="paymentAmount-label"
          >
            Zahlungssumme
            <span
              aria-hidden="true"
              class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
            >
               
              *
            </span>
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedEnd css-o9k5xi-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
              id="paymentAmount"
              required=""
              type="number"
              value="123"
            />
            <div
              class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 css-1pnmrwp-MuiTypography-root"
              >
                €
              </p>
            </div>
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-14lo706"
              >
                <span>
                  Zahlungssumme
                   
                  *
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
      >
        <div
          class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="true"
            for="paymentDate"
            id="paymentDate-label"
          >
            Zahlungsempfang
            <span
              aria-hidden="true"
              class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
            >
               
              *
            </span>
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              autocomplete="off"
              class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id="paymentDate"
              inputmode="text"
              placeholder="TT.MM.JJJJ"
              required=""
              type="text"
              value="28.05.2023"
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-14lo706"
              >
                <span>
                  Zahlungsempfang
                   
                  *
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
    </div>
  </form>
</div>
`;
